dll/win32/iphlpapi/iphlpapi_main.c

```diff
@@ -2620,12 +2621,113 @@ SetIpForwardEntryToStack(PMIB_IPFORWARDROW pRoute)
 }
 
 DWORD WINAPI
-NhGetInterfaceNameFromDeviceGuid(DWORD dwUnknown1,
-                                 DWORD dwUnknown2,
-                                 DWORD dwUnknown3,
+NhGetInterfaceNameFromDeviceGuid(GUID * pInterfaceGUID,
+                                 PWCHAR pInterfaceName,
+                                 PULONG pOutBufLen,
                                  DWORD dwUnknown4,
                                  DWORD dwUnknown5)
 {
-    FIXME("NhGetInterfaceNameFromDeviceGuid() stub\n");
-    return 0L;
+    UNICODE_STRING GuidString;
+    WCHAR szKeyName[2*MAX_PATH];
+    HKEY hKey;
+    DWORD result, type;
+
+    if (pInterfaceGUID == NULL || pOutBufLen == NULL)
+    {
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    if (RtlStringFromGUID(pInterfaceGUID, &GuidString) != STATUS_SUCCESS)
+    {
+        // failed to convert guid to string
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    if (StringCbPrintfW(szKeyName, sizeof(szKeyName), L"SYSTEM\\CurrentControlSet\\Control\\Network\\{4D36E972-E325-11CE-BFC1-08002BE10318}\\%s\\Connection", GuidString.Buffer) != S_OK)
+    {
+        // key name is too long
+        RtlFreeUnicodeString(&GuidString);
+        return ERROR_BUFFER_OVERFLOW;
+    }
+    RtlFreeUnicodeString(&GuidString);
+
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE, szKeyName, 0, KEY_READ, &hKey) != ERROR_SUCCESS) 
+    {
+        // failed to find adapter entry
+        return ERROR_FILE_NOT_FOUND;
```

Comment:
Why not propagate RegOpenKeyEx's return value?

Reply:
I did not tested yet return values of these function, the only thing I know it returns 0 on success, so I decided to return just error codes.

Maybe it uses SetLastError, but this needs to be tested.